modules/substrate/src/lib.rs

```diff
@@ -433,3 +464,65 @@ impl<T: Trait> BridgeStorage for PalletStorage<T> {
 		<NextScheduledChange<T>>::insert(signal_hash, next_change)
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::mock::{helpers::test_header, run_test, TestRuntime};
+	use frame_support::{assert_noop, assert_ok};
+
+	#[test]
+	fn parse_finalized_storage_proof_rejects_proof_on_unknown_header() {
+		run_test(|| {
+			assert_noop!(
+				Module::<TestRuntime>::parse_finalized_storage_proof(
+					Default::default(),
+					StorageProof::new(vec![]),
+					|_| (),
+				),
+				Error::<TestRuntime>::UnknownHeader,
+			);
+		});
+	}
+
+	#[test]
+	fn parse_finalized_storage_proof_rejects_proof_on_unfinalized_header() {
+		run_test(|| {
+			let mut storage = PalletStorage::<TestRuntime>::new();
+			let header = test_header(1);
+			storage.write_header(&ImportedHeader {
+				header: header.clone(),
+				requires_justification: false,
+				is_finalized: false,
+				signal_hash: None,
+			});
+
+			assert_noop!(
+				Module::<TestRuntime>::parse_finalized_storage_proof(header.hash(), StorageProof::new(vec![]), |_| (),),
+				Error::<TestRuntime>::UnfinalizedHeader,
+			);
+		});
+	}
+
+	#[test]
+	fn parse_finalized_storage_accepts_valid_proof() {
+		run_test(|| {
+			let mut storage = PalletStorage::<TestRuntime>::new();
+			let (state_root, storage_proof) = storage_proof::tests::craft_valid_storage_proof();
+			let mut header = test_header(1);
+			header.set_state_root(state_root);
+
+			storage.write_header(&ImportedHeader {
+				header: header.clone(),
+				requires_justification: false,
+				is_finalized: true,
+				signal_hash: None,
+			});
```

Comment:
You could probably add a `DerefMut` implementation to `ImportedHeader` and do something like:

```rust
let mut header = unfinalized_header(1);
header.is_finalized = true;
header.set_storage_root(state_root);
storage.write_header(&header);
```

No big deal though, current code is fine